File: src/net/dinkla/raytracer/materials/SVMatte.java
package net.dinkla.raytracer.materials;

import net.dinkla.raytracer.brdf.SVLambertian;
import net.dinkla.raytracer.colors.ColorAccumulator;
import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.lights.AreaLight;
import net.dinkla.raytracer.lights.Light;
import net.dinkla.raytracer.math.Ray;
import net.dinkla.raytracer.math.Vector3D;
import net.dinkla.raytracer.textures.Texture;
import net.dinkla.raytracer.worlds.World;

import java.util.List;

public class SVMatte extends Material {

    public SVLambertian ambientBrdf;
    public SVLambertian diffuseBrdf;

    public SVMatte() {
        ambientBrdf = new SVLambertian();
        diffuseBrdf = new SVLambertian();
        setKa(0.25);
        setKd(0.75);
        setCd(null);
    }

    public SVMatte(final Texture color, final double ka, final double kd) {
        ambientBrdf = new SVLambertian();
        diffuseBrdf = new SVLambertian();
        setKa(ka);
        setKd(kd);
        setCd(color);
    }

    public void setKa(double ka) {
        ambientBrdf.setKd(ka);
    }

    public void setKd(double kd) {
        diffuseBrdf.setKd(kd);
    }

    public void setCd(final Texture cd) {
        ambientBrdf.setCd(cd);
        diffuseBrdf.setCd(cd);
    }

    @Override
    public Color shade(World world, Shade sr) {
        Vector3D wo = sr.ray.getD().negate();
        Color L = getAmbientColor(world, sr, wo);
        for (Light light : world.getLights()) {
            Vector3D wi = light.getDirection(sr);
            double nDotWi = wi.dot(sr.getNormal());
            if (nDotWi > 0) {
                boolean inShadow = false;
                if (light.getShadows()) {
                    Ray shadowRay = new Ray(sr.getHitPoint(), wi);
                    inShadow = light.inShadow(world, shadowRay, sr);
                }
                if (!inShadow) {
                    Color f = diffuseBrdf.f(sr, wo, wi);
                    Color l = light.L(world, sr);
                    Color flndotwi = f.mult(l).mult(nDotWi);
                    L = L.plus(flndotwi);
                }
            }
        }
        return L;
    }

    @Override
    public Color areaLightShade(World world, Shade sr) {
        Vector3D wo = sr.ray.getD().negate();
        Color L = getAmbientColor(world, sr, wo);
        ColorAccumulator S = new ColorAccumulator();
        for (Light light1 : world.getLights()) {
            if (light1 instanceof AreaLight) {
                AreaLight light = (AreaLight) light1;
                List<AreaLight.Sample> ls = light.getSamples(sr);
                for (AreaLight.Sample sample : ls) {
                    double nDotWi = sample.wi.dot(sr.getNormal());
                    if (nDotWi > 0) {
                        boolean inShadow = false;
                        if (light.getShadows()) {
                            Ray shadowRay = new Ray(sr.getHitPoint(), sample.wi);
                            inShadow = light.inShadow(world, shadowRay, sr, sample);
                        }
                        if (!inShadow) {
                            Color f = diffuseBrdf.f(sr, wo, sample.wi);
                            Color l = light.L(world, sr, sample);
                            Color flndotwi = f.mult(l).mult(nDotWi);
                            // TODO: hier ist der Unterschied zu shade()
                            double f1 = light.G(sr, sample) / light.pdf(sr);
                            Color T = flndotwi.mult(f1);
                            S.plus(T);
                        }
                    }
                }
            }
        }
        L = L.plus(S.getAverage());
        return L;
    }

    protected Color getAmbientColor(World world, Shade sr, Vector3D wo) {
        Color c1 = ambientBrdf.rho(sr, wo);
        Color c2 = world.getAmbientLight().L(world, sr);
        Color L = c1.mult(c2);
        return L;
    }

    @Override
    public Color getLe(Shade sr) {
        return diffuseBrdf.rho(sr, null);
    }
}

File: src/jvmMain/kotlin/net/dinkla/raytracer/materials/SVMatte.kt
package net.dinkla.raytracer.materials

import net.dinkla.raytracer.brdf.SVLambertian
import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.colors.ColorAccumulator
import net.dinkla.raytracer.hits.IShade
import net.dinkla.raytracer.lights.AreaLight
import net.dinkla.raytracer.math.Ray
import net.dinkla.raytracer.math.Vector3D
import net.dinkla.raytracer.textures.Texture
import net.dinkla.raytracer.world.IWorld

open class SVMatte : IMaterial {

    var ambientBrdf: SVLambertian
    var diffuseBrdf: SVLambertian

    constructor() {
        ambientBrdf = SVLambertian()
        diffuseBrdf = SVLambertian()
        setKa(0.25)
        setKd(0.75)
        setCd(null)
    }

    constructor(color: Texture, ka: Double, kd: Double) {
        ambientBrdf = SVLambertian()
        diffuseBrdf = SVLambertian()
        setKa(ka)
        setKd(kd)
        setCd(color)
    }

    fun setKa(ka: Double) {
        ambientBrdf.kd = ka
    }

    fun setKd(kd: Double) {
        diffuseBrdf.kd = kd
    }

    fun setCd(cd: Texture?) {
        ambientBrdf.cd = cd
        diffuseBrdf.cd = cd
    }

    override fun shade(world: IWorld, sr: IShade): Color {
        val wo = -sr.ray.direction
        var L = getAmbientColor(world, sr, wo)
        for (light in world.lights) {
            val wi = light.getDirection(sr)
            val nDotWi = wi.dot(sr.normal)
            if (nDotWi > 0) {
                var inShadow = false
                if (light.shadows) {
                    val shadowRay = Ray(sr.hitPoint, wi)
                    inShadow = light.inShadow(world, shadowRay, sr)
                }
                if (!inShadow) {
                    val f = diffuseBrdf.f(sr, wo, wi)
                    val l = light.L(world, sr)
                    val flndotwi = f.times(l).times(nDotWi)
                    L = L.plus(flndotwi)
                }
            }
        }
        return L
    }

    override fun areaLightShade(world: IWorld, sr: IShade): Color {
        val wo = -sr.ray.direction
        var L = getAmbientColor(world, sr, wo)
        val S = ColorAccumulator()
        for (light1 in world.lights) {
            if (light1 is AreaLight) {
                val ls = light1.getSamples(sr)
                for (sample in ls) {
                    val nDotWi = sample.wi!!.dot(sr.normal)
                    if (nDotWi > 0) {
                        var inShadow = false
                        if (light1.shadows) {
                            val shadowRay = Ray(sr.hitPoint, sample.wi!!)
                            inShadow = light1.inShadow(world, shadowRay, sr, sample)
                        }
                        if (!inShadow) {
                            val f = diffuseBrdf.f(sr, wo, sample.wi!!)
                            val l = light1.L(world, sr, sample)
                            val flndotwi = f.times(l).times(nDotWi)
                            // TODO: hier ist der Unterschied zu shade()
                            val f1 = light1.G(sr, sample) / light1.pdf(sr)
                            val T = flndotwi.times(f1)
                            S.plus(T)
                        }
                    }
                }
            }
        }
        L = L.plus(S.average)
        return L
    }

    protected fun getAmbientColor(world: IWorld, sr: IShade, wo: Vector3D): Color {
        val c1 = ambientBrdf.rho(sr, wo)
        val c2 = world.ambientLight.L(world, sr)
        return c1.times(c2)
    }

    override fun getLe(sr: IShade): Color {
        return diffuseBrdf.rho(sr, null!!)
    }
}
